### PR TITLE
[Feature] Add visibility prediction head

### PR DESCRIPTION
**Pull request by Billccx**
## Motivation
Enable top-down methods to predict keypoints visibility

## Modification
1. add a new head `vis_head` in hybrid_heads      (mmpose/mmpose/models/heads/hybrid_heads/vis_head.py)
2. add a new TTA function `flip_visibility`     (mmpose/mmpose/models/utils/tta.py)
3. add a new key `transformed_keypoints_visible` in `label_mapping_table` (mmpose/mmpose/datasets/transforms/formatting.py)
4. support `keypoints_visible` in `GenerateTarget` (mmpose/mmpose/datasets/transforms/common_transforms.py)
5. <s>add a new loss function `BCEWithLogitsLoss` (mmpose/mmpose/models/losses/classification_loss.py)</s>
6.  refine `BCELoss` module: add parameter `with_logits` to decide whether to use BCEWithLogitsLoss  (mmpose/mmpose/models/losses/classification_loss.py)
7. add unittest for `vis_head`  (mmpose/tests/test_models/test_heads/test_hybrid_heads/test_vis_head.py)

## Use cases (Optional)

We can use `vis_head` to wrap a keypoints prediction head, getting the keypoints prediction results with keypoints visibility.

For example, in config file:

```
head=dict(
  type='VisPredictHead',
  loss=dict(type='BCELoss', use_target_weight=False, with_logits=True),
  use_sigmoid=False,
  pose_cfg=dict(
    type='HeatmapHead',
    in_channels=32,
    out_channels=17,
    deconv_out_channels=None,
    loss=dict(type='KeypointMSELoss', use_target_weight=True),
    decoder=codec
  )
)
```

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpose/models/heads/hybrid_heads/vis_head.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        else:
+            pose_pred_instances = pose_preds
+            pose_pred_fields = None
+
+        batch_vis_np = to_numpy(batch_vis, unzip=True)
+
+        assert len(pose_pred_instances) == len(batch_vis_np)
+        for index in range(len(pose_pred_instances)):
+            pose_pred_instances[index].keypoint_visibility = batch_vis_np[
+                index]
```
Comment (Ben-Louis) [inline mmpose/models/heads/hybrid_heads/vis_head.py:111]: Why has a loop been utilized here? BTW, the visibility representation in 'pred_instances' is denoted by 'keypoint_scores'

Comment (Billccx) [reply mmpose/models/heads/hybrid_heads/vis_head.py:111]: > Why has a loop been utilized here? BTW, the visibility representation in 'pred_instances' is denoted by 'keypoint_scores'

Thanks for your advise. I will rewrite this part soon, sorry for the bad code.

Comment (Ben-Louis) [reply mmpose/models/heads/hybrid_heads/vis_head.py:111]: I'm sorry, my tone may have led to a misunderstanding. I was just unclear whether 'pose_pred_instances' is a tuple or InstanceData, so I wanted to ask if there is a possibility of not using a loop. Take it easy~

Comment (Billccx) [reply mmpose/models/heads/hybrid_heads/vis_head.py]: > I'm sorry, my tone may have led to a misunderstanding. I was just unclear whether 'pose_pred_instances' is a tuple or InstanceData, so I wanted to ask if there is a possibility of not using a loop. Take it easy~

Got it! Sorry for the late reply. 

From my perspective, 'pose_pred_instances' is an 'InstanceList'? 

> https://github.com/open-mmlab/mmpose/blob/2c4a60e357c7f68f3388f0fc4ddcb404674b81ae/mmpose/models/heads/base_head.py#L78-L81

> https://github.com/open-mmlab/mmpose/blob/2c4a60e357c7f68f3388f0fc4ddcb404674b81ae/mmpose/models/heads/heatmap_heads/heatmap_head.py#L235-L241

Besides that, I have refined the code. If I still get something wrong, please let me know. Thanks.

File: mmpose/models/heads/hybrid_heads/vis_head.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            _batch_vis_flip = flip_visibility(
+                self.vis_forward(_feats_flip), flip_indices=flip_indices)
+            batch_vis = (_batch_vis + _batch_vis_flip) * 0.5
+        else:
+            batch_vis = self.vis_forward(feats)  # (B, K, D)
+
+        batch_vis.unsqueeze_(dim=1)  # (B, N, K, D)
+
+        if not self.use_sigmoid:
+            batch_vis = torch.sigmoid(batch_vis)
```
Comment (Ben-Louis) [inline mmpose/models/heads/hybrid_heads/vis_head.py:171]: Why is `sigmoid` applied to `batch_vis` when `use_sigmoid` is `False`?

Comment (Billccx) [reply mmpose/models/heads/hybrid_heads/vis_head.py:171]: I think the value of visibility prediction should be between 0 and 1? If `use_sigmoid` is `False`, there won't be sigmoid in `vis_forward`.

Comment (Ben-Louis) [reply mmpose/models/heads/hybrid_heads/vis_head.py:171]: If the sigmoid function is always applied to the head outputs, what is the role of the `use_sigmoid` argument?

Comment (Tau-J) [reply mmpose/models/heads/hybrid_heads/vis_head.py:171]: > Why is `sigmoid` applied to `batch_vis` when `use_sigmoid` is `False`?

@Ben-Louis When `use_sigmoid` is True, a `nn.Sigmoid()` will be in the `vis_head`, namely, the `batch_vis` is between 0~1. Otherwise, we need to do sigmoid manually when `use_sigmoid` is False.

Comment (Tau-J) [reply mmpose/models/heads/hybrid_heads/vis_head.py:171]: Using `nn.Sigmoid()` and `BCELoss` separately is not officially recommended by Pytorch, a better way is to use `BCEWithLogitsLoss` and do sigmoid manually when a 0-1 score is needed. So I suggest him to set a `use_sigmoid` to keep the flexity.

Comment (Ben-Louis) [reply mmpose/models/heads/hybrid_heads/vis_head.py:171]: @Tau-J, thanks for the clarification. This can indeed be puzzling, perhaps we could add some comments for clarity? Also, focusing on a single recommended practice could be beneficial. BTW, in MMDetection, this parameter is generally integrated in loss modules like `mmdet.CrossEntropyLoss`. We might consider refining the `BCELoss` module with this parameter, rather than introducing a completely new loss module.

**Review by Tau-J — APPROVED** [review]
